query.go

```diff
@@ -70,10 +74,207 @@ func (q *dhtQuery) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, e
 	return runner.Run(ctx, peers)
 }
 
+type stagingQueue struct {
+	dialFn func(context.Context, peer.ID) error
+	ctx    context.Context
+	// TODO maxParallelism int
+
+	parLk         sync.Mutex
+	parallelism   int
+	scalingFactor float64
+
+	in          *queue.ChanQueue
+	completedCh chan peer.ID
+
+	dieCh    chan struct{}
+	notifyCh chan struct{}
+
+	retChs   []chan dialResult
+	retChIdx int32
+}
+
+var StagingQueueMinParallelism = 6
+var StagingQueueMaxIdle = 5 * time.Second
+
+var ErrContextClosed = errors.New("context closed")
+
+func newStagingQueue(ctx context.Context, in *queue.ChanQueue, dialFn func(context.Context, peer.ID) error, nConsumers int) *stagingQueue {
+	sq := &stagingQueue{
+		dialFn:        dialFn,
+		ctx:           ctx,
+		scalingFactor: 1.5,
+		parallelism:   StagingQueueMinParallelism,
+		in:            in,
+
+		// TODO: we should really re-sort dials as they come back based on XORKeySpace, instead of queuing them in the
+		// order they come back.
+		completedCh: make(chan peer.ID),
```

Comment:
We could use another priority queue. However, we'll need some other backpressure mechanism, some kind of ticket system?